domain/secret/state/state.go

```diff
@@ -1883,3 +1949,268 @@ ON CONFLICT(secret_id) DO UPDATE SET
+AND    sp.subject_uuid = $secretPermission.subject_uuid
+AND    (sp.subject_type_id <> $secretPermission.subject_type_id
+        OR sp.scope_uuid <> $secretPermission.scope_uuid
+        OR sp.scope_type_id <> $secretPermission.scope_type_id)
+`
+
+	checkInvariantStmt, err := st.Prepare(checkInvariantQuery, secretPermission{}, secretID{})
+	if err != nil {
+		return errors.Trace(err)
+	}
+
+	err = db.Txn(ctx, func(ctx context.Context, tx *sqlair.TX) error {
+		perm := secretPermission{
+			SecretID: uri.ID,
+			RoleID:   params.RoleID,
+		}
+		if isLocal, err := st.checkExistsIfLocal(ctx, tx, uri); err != nil {
+			return errors.Trace(err)
+		} else if !isLocal {
+			// Should never happen.
+			return secreterrors.SecretIsNotLocal
+		}
+
+		// Look up the UUID of the subject.
+		result := entityRef{}
+		err = tx.Query(ctx, selectSubjectUUIDStmt, selectSubjectQueryParams).Get(&result)
+		if err != nil {
+			if errors.Is(err, sqlair.ErrNoRows) {
+				return fmt.Errorf("%#v %q not found%w", params.SubjectTypeID, params.SubjectID, errors.Hide(subjectNotFoundError))
+			} else {
+				subject := params.SubjectID
+				if params.SubjectTypeID == domainsecret.SubjectModel {
+					subject = "model"
+				}
+				return errors.Annotatef(err, "looking up secret grant subject UUID for %q", subject)
+			}
+		}
+		perm.SubjectUUID = result.UUID
+		perm.SubjectTypeID = params.SubjectTypeID
+
+		// Look up the UUID of the access scope entity.
+		result = entityRef{}
+		err = tx.Query(ctx, selectScopeUUIDStmt, selectScopeQueryParams).Get(&result)
+		if err != nil {
+			if errors.Is(err, sqlair.ErrNoRows) {
+				return fmt.Errorf("%#v %q not found%w", params.ScopeTypeID, params.ScopeID, errors.Hide(scopeNotFoundError))
+			} else {
+				scope := params.ScopeID
+				if params.ScopeTypeID == domainsecret.ScopeModel {
+					scope = "model"
+				}
+				return errors.Annotatef(err, "looking up secret grant scope UUID for %q", scope)
+			}
+		}
+		perm.ScopeUUID = result.UUID
+		perm.ScopeTypeID = params.ScopeTypeID
+
+		// Check that the access scope or subject type is not changing.
+		id := secretID{}
+		err = tx.Query(ctx, checkInvariantStmt, perm).Get(&id)
+		if err == nil {
+			// Should never happen.
+			return secreterrors.InvalidSecretPermissionChange
+		} else if !errors.Is(err, sqlair.ErrNoRows) {
+			return errors.Annotatef(err, "checking duplicate permission record for secret %q", uri)
+		}
+
+		return st.grantAccess(ctx, tx, perm)
+	})
+	return errors.Trace(domain.CoerceError(err))
+}
+
+func (st State) grantAccess(ctx context.Context, tx *sqlair.TX, perm secretPermission) error {
+	insertQuery := `
+INSERT INTO secret_permission (*)
+VALUES ($secretPermission.*)
+ON CONFLICT(secret_id, subject_uuid) DO UPDATE SET
+    role_id=excluded.role_id,
+    -- These are needed to fire the immutable trigger.
+    subject_type_id=excluded.subject_type_id,
+    scope_type_id=excluded.scope_type_id,
+    scope_uuid=excluded.scope_uuid
```

Comment:
Can we remove triggersForImmutableTableUpdates and reuse the `triggersForImmutableTable` by introducing an operation enum as a drive by?

Reply:
It's only a few lines of code. It's simpler just to have a couple of functions. The overhead of parametrising it doesn't really pay its way